### PR TITLE
Generics analysis

### PR DESCRIPTION
**Pull request by mtrive**
This is an experimental "Generics" view which reports all Generic types instantiations. It works similarly to the Code view, except the instantiations are grouped by Generic type.

generics-auditing

In the future (once the UI allows for that), this will change so that instantiations will also be grouped depending on whether the Generic types parameters are references or value types. For example:

- `GenericType<>`
  - `GenericType<object>`
    - all the reference type instantiations (they together only produce 1 type in IL2CPP)
  - `GenericType<float>`
  - `GenericType<int>`

### CODE REVIEW

File: Editor/Auditors/ScriptAuditor.cs [hunk 1]
```diff
@@ -56,6 +67,7 @@ public IEnumerable<ProblemDescriptor> GetDescriptors()
         public IEnumerable<IssueLayout> GetLayouts()
         {
             yield return k_IssueLayout;
+            yield return k_GenericIssueLayout;
```
Comment (berniegp) [inline Editor/Auditors/ScriptAuditor.cs:70]: Nitpick really, but in other parts of the code this is called "Generics" with an s. Might be nice to name it consistently.

File: Editor/InstructionAnalyzers/GenericTypeInstantiationAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+using Mono.Cecil.Cil;
+using Unity.ProjectAuditor.Editor.CodeAnalysis;
+
+namespace Unity.ProjectAuditor.Editor.InstructionAnalyzers
+{
+    public class GenericTypeInstantiationAnalyzer : IInstructionAnalyzer
+    {
+        const int k_FirstDescriptorId = 500000;
+
+        Dictionary<string, ProblemDescriptor> m_GenericDescriptors = new Dictionary<string, ProblemDescriptor>();
```
Comment (berniegp) [inline Editor/InstructionAnalyzers/GenericTypeInstantiationAnalyzer.cs:12]: Nitpick really, but in other parts of the code this is called "Generics" with an s. Might be nice to name it consistently.

**Review by berniegp — APPROVED** [review]
Approved with minor comments.

I don't fully understand the way `GenericTypeInstantiationAnalyzer` works, but nothing stands out to me.

Good job on the overall design btw. I'm surprised how little code was required to build this feature.

**Comment by mtrive** [comment]
> Approved with minor comments.
> 
> I don't fully understand the way `GenericTypeInstantiationAnalyzer` works, but nothing stands out to me.
> 
> Good job on the overall design btw. I'm surprised how little code was required to build this feature.

Just a few months ago it would not have been this easy. It took a lot of refactor but now I think the current design works quite well for this Cecil-based analysis.

I agree with the naming consistency comments, I will probably removing all the 's' from the IssueCategory class enums, in a separate PR though.

**Pull request merged**